internal/controller/pkg/revision/establisher.go

```diff
@@ -73,6 +80,27 @@ type currentDesired struct {
 func (e *APIEstablisher) Establish(ctx context.Context, objs []runtime.Object, parent resource.Object, control bool) ([]xpv1.TypedReference, error) { // nolint:gocyclo
 	allObjs := []currentDesired{}
 	resourceRefs := []xpv1.TypedReference{}
+	var rev v1.PackageRevision
+	switch r := parent.(type) {
+	case *v1.ProviderRevision:
+		rev = r
+	case *v1.ConfigurationRevision:
+		rev = r
+	default:
+		return nil, errors.New("type of parent cannot be other than providerrevision or configurationrevision")
+	}
+	var webhookTLSCert []byte
+	if rev.GetWebhookTLSSecretName() != nil {
```

Comment:
What this does right now is to deploy configurations together with CRDs and then move on to deploying `Deployment` and `Service`. In the upgrade scenario, there is an interval where the webhook config is set to the new revision's `Service` that's yet to be deployed, hence all requests would fail in that interval. The best way to do that would've been to update the webhook configs after the new `Service` is deployed.

I looked into that briefly and couldn't see a way to do it without a lot of surgery but will take another look.

Reply:
It seems like I can access the package content with `pkg.GetObjects()` in the `Post` hook of provider hooks, so I can move application of webhook configs there. But we'd still need to fetch the TLS secret and inject the CA bundle into CRDs in the establisher. @hasheddan thoughts?

Reply:
No, I can't do that without passing the array since the `pkg` we get seems to need to be `*parser.Package` object. Though I can try some surgery if you think it'd be inline with the general flow.

Reply:
@muvaf with the current strategy, it sounds like there is a period where _all_ requests are rejected. With the proposed change in sounds like there is a period where some requests that would be rejected would not be (i.e. while waiting for the webhooks to be put into place). In practice, I doubt the latter would be a frequent occurrence, but I would prefer to bias towards the former (fail until we have the setup that is expected).

Reply:
Sounds good. Thanks for the input!